<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Population Dynamics - The Little Book of Computational Biology</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="intro.html">Introduction</a></li><li class="chapter-item expanded "><a href="DNA.html"><strong aria-hidden="true">1.</strong> DNA</a></li><li class="chapter-item expanded "><a href="cells.html"><strong aria-hidden="true">2.</strong> Protein and Cells</a></li><li class="chapter-item expanded "><a href="transcription-networks.html"><strong aria-hidden="true">3.</strong> Transcription Networks</a></li><li class="chapter-item expanded "><a href="sequence-matching.html"><strong aria-hidden="true">4.</strong> Sequence Matching</a></li><li class="chapter-item expanded "><a href="hmm.html"><strong aria-hidden="true">5.</strong> Hidden Markov Models</a></li><li class="chapter-item expanded "><a href="neural.html"><strong aria-hidden="true">6.</strong> Neural Applications</a></li><li class="chapter-item expanded "><a href="population-dynamics.html" class="active"><strong aria-hidden="true">7.</strong> Population Dynamics</a></li><li class="chapter-item expanded "><a href="evo-trees.html"><strong aria-hidden="true">8.</strong> Evolutionary Trees</a></li><li class="chapter-item expanded "><a href="boolean-networks.html"><strong aria-hidden="true">9.</strong> Boolean Networks</a></li><li class="chapter-item expanded "><a href="epigenetics.html"><strong aria-hidden="true">10.</strong> Epigenetics</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">The Little Book of Computational Biology</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/jackdarlison/the-little-book-of-computational-biology" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="population-dynamics"><a class="header" href="#population-dynamics">Population Dynamics</a></h1>
<p>Population dynamics looks at how population levels vary of time given certain factors. Formed by combining biostatistics and Mendelian inheritance</p>
<p>Prior to Mendel it was thought characteristic were blended, <strong>Mendelian Inheritance</strong> is the idea that the parents characteristics are independent and children independent parts of both not a mix. </p>
<p>Biostatistics cover the statistics population dynamics:</p>
<ul>
<li>Wright found:
<ul>
<li>Inbreeding gave more variation</li>
<li>Small isolated populations led to more variation towards a local optima</li>
</ul>
</li>
<li>Fisher found: 
<ul>
<li>Large populations don't need to worry about local optima</li>
<li>Sufficiently large populations will always move &quot;uphill&quot;</li>
</ul>
</li>
<li>Haldne found:
<ul>
<li>certain features are inherited together (e.g. genetic linkage)</li>
<li>Provided evidence for inheritance via chromosomes</li>
</ul>
</li>
</ul>
<p>Modern synthesis of evolution brings together:</p>
<ul>
<li>Natural selection</li>
<li>Population genetics</li>
<li>Medelian genetics</li>
</ul>
<p>We know that:</p>
<ul>
<li>Genes can mutate</li>
<li>Alleles are exchanged during sexual reproduction</li>
<li>Evolutionary fitness measures the long term number of offspring</li>
<li>Epistasis, genes interact</li>
<li>Linkage disequilibrium</li>
<li>Generations overlap</li>
</ul>
<h3 id="fisher-wright-model"><a class="header" href="#fisher-wright-model">Fisher Wright model</a></h3>
<p>The FW model looks at how a single mutation can take over in a population.</p>
<p>This model assumes:</p>
<ul>
<li>genes can be changed from one allele to another via mutation</li>
<li>Alleles are exchanged during sexual reproduction</li>
<li>Fitness is a measure of the expected number of offspring</li>
<li>no epistasis</li>
<li>linkage equilibrium</li>
<li>generations do not overlap</li>
</ul>
<p>To make models with FW we make further assumptions:</p>
<ul>
<li>No gender</li>
<li>Haploid (one copy of genes)</li>
<li>Two states normal (P) and mutant (Q)</li>
<li>Q has selective advantage of 1+s</li>
<li>P mutates to Q at rate m</li>
<li>Q mutates to P at rate n</li>
<li>constant population size (N) (for infinite populations we can approximate using differential equation)</li>
</ul>
<p>Each generation follows:</p>
<ul>
<li>individuals produce seeds</li>
<li>seeds mutate according to m and n</li>
<li>sample N of the seeds</li>
</ul>
<p>In reality we don't use exact numbers for seeds, only the proportion of mutants compared to normal, the model:</p>
<ul>
<li>Find the proportion of mutants after seeding: \( p_s = \frac{(1+s)x}{N+sx} \) where x is the number of mutants in this generation</li>
<li>Mutate the seeds and find the new proportion of mutants: \( p_{sm} = (1-n)p_s + m(1 - p_s) \)</li>
<li>Sample N individuals, using a binomial with the proportion of mutants: \( P(x(t+1) = x) = \binom{N}{x} p^x_{sm}(1-p_{sm})^{N-x} \)</li>
</ul>
<p>If we presume that the population is infinite we can just use proportions instead of numbers: \[ p_x(t+1) = \frac{(1-n)(1+s)p_x(t) + m(1 - p_x(t))}{1 + sp_x(t)} \]</p>
<p>Which can be approximated by a differential equation if the effects of selection and mutation are small: \[ \frac{dp_x}{dt} \approx p_x(t+1) - p_x(t) \]</p>
<p>The problem with this is that if mutants are fitter, they will grow exponentially at the start. ODEs have much more rapid take over times than finite population models if low mutation rates</p>
<p>Finite models are stochastic, where as infinite populations are deterministic.</p>
<h3 id="markov-models"><a class="header" href="#markov-models">Markov Models</a></h3>
<p>Evolving populations can be modelled using a Markov chains. </p>
<p>To do this we define:</p>
<ul>
<li>A transition matrix \( W_{x', x} \) where each cell represents the probability from x to x' mutants.</li>
<li>The probability of x mutants are updated with \( p(t+1) = Wp(t) \)</li>
<li>The start prob for no mutants:  \( p(t=0) = [1, 0, 0, ...] \)</li>
</ul>
<p>Thus, \(p(t) = W^tp(0) \)</p>
<p>This model creates a probability distribution over the number of mutants for each time step. </p>
<p>The transition matrix is given: </p>
<p>\[ W_{x', x} = \binom{N}{x'} p_sm(x)^{x'}(1-p_{sm}(x)^{N-x'}) \]</p>
<p>Where:</p>
<p>\[ p_{sm}(x) = \frac{(1-n)(1+s)x + m(N-x)}{N+sx} \]</p>
<p>Markov models provide an exact analysis of population dynamics. However, as the  matrix is the size of the states + 1 squared, this model is infeasible for large populations.</p>
<h3 id="stochastic-differential-equations"><a class="header" href="#stochastic-differential-equations">Stochastic Differential Equations</a></h3>
<p>To model the randomness inherent in population dynamics, stochastic differential equations are used. These work best with weak selection and small mutation rates</p>
<p>Presume a random walk (±1). The expectation of distance travelled is zero, as well as the expectation of number of steps in total. However, the expectation of the squared distance is somehow equal √N </p>
<p>With random time steps dt and for a walk with distance x, the change is distance is measured:</p>
<p>\[ dx = adt + bdW(t) \]</p>
<p>where:</p>
<ul>
<li>a is the drift, i.e the expected change</li>
<li>b is the variance, due to random processes</li>
<li>W(t) is a random process drawing from a normal distribution with mean 0 and variance dt: i.e.  N(0, dt)</li>
</ul>
<p>This is closely related to a Maxwell-Boltzmann distribution.</p>
<p>Applying this to evolution:</p>
<ul>
<li>Time steps are generations</li>
<li>Distance is the number of mutants in the population</li>
<li>dx is the change in mutants in each time step</li>
</ul>
<p>We get:</p>
<p>\[ X(t+dt) = X(t) + a(X(t))dt + b(X(t))dW(t) \]</p>
<p>Where:</p>
<ul>
<li>Expected change / drift: a(X(t)) = (p_{sm}(X(t)) - X(t))dt</li>
<li>Variance: b(X(t)) = \frac{p_{sm}(X(t))(1-p_{sm}(X(t)))}{N}dt</li>
</ul>
<p>If s, m, and n are small then \( p_{sm} \approx X(t) \) and can be substituted in variance.</p>
<p>We get a stochastic differential equation or SDE:</p>
<p>\[ dX(t) = a(X(t))dt + b(X(t))dW(t) \]</p>
<p>This is known as an Itô process which can be solved. There is no closed from solution for the differential of W(t) as this will also be stochastic. We can however solve the probability distribution: \( f(x, t)dx = P(x \le X(t) \lt x + dx) \)</p>
<p>The solution to this is given by the <strong>Fokker-Planck</strong> equation (i.e. Forward Kolmogorov):</p>
<p>\[ \frac{\partial f(x, t)}{\partial t} = -\frac{a(x)\partial f(x, t)}{\partial x} + \frac{1}{2} \frac{\partial^2(b^2(x)f(x, t))}{\partial x^2} \]</p>
<p>However diffusion doesn't go backwards so we need to the Backward Kolmogorov:</p>
<p>\[ \frac{\partial f(x, -t)}{\partial t} = -a(x)\frac{\partial f(x, -t)}{\partial x} + \frac{1}{2} b^2(x)\frac{\partial^2(f(x, -t))}{\partial x^2} \]</p>
<p>This can be used to work out how long it might take to get to an absorbing state and how likely this state may be</p>
<p>Fokker-Planck shifts the SDE into a PDE. This is hard to solve directly but can use numerical methods. This is almost identical to Markov Chain analysis and can scale well to large numbers</p>
<h3 id="steady-state-distribution"><a class="header" href="#steady-state-distribution">Steady State Distribution</a></h3>
<p>Solving the PDE is hard in the previous model. We can approximate this by solving for when it equals 0, i.e. when the distribution stops changing at t tends to ∞. This can be done as solving for t tends to ∞ does not depend on t. </p>
<p>It is solved using the equation:</p>
<p>\[ f_{eq}(x) = \frac{x^{2Nm-1}(1-x)^{Nn-1}e^{2Nsx}}{\text{Betaln}(2Nm, 2Nn)\text{Hyp1f1}(2Nm; 2N(m + n); 2Nsx)} \]</p>
<p>This is a valid approximation as evolution is slow, which means the steady state is often seen. </p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="neural.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="evo-trees.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="neural.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="evo-trees.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>

        <!-- Livereload script (if served using the cli tool) -->
        <script>
            const wsProtocol = location.protocol === 'https:' ? 'wss:' : 'ws:';
            const wsAddress = wsProtocol + "//" + location.host + "/" + "__livereload";
            const socket = new WebSocket(wsAddress);
            socket.onmessage = function (event) {
                if (event.data === "reload") {
                    socket.close();
                    location.reload();
                }
            };

            window.onbeforeunload = function() {
                socket.close();
            }
        </script>



        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
